File: leet_BST_0938_rangesum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        return [
            # self.BFS,
            # self.DFS
            self.Inorder,
        ][ 0 ]( root, low, high )

    def BFS(self, root: Optional[TreeNode], low: int, high: int) -> int:
        Q = deque()
        Q.append(root)
        res = []
        while Q:
            node = Q.popleft()
            if low <= node.val <= high:
                res.append(node.val)
            if node.left:
                Q.append(node.left)
            if node.right:
                Q.append(node.right)
        return sum(res)

    def Inorder(self, root: Optional[TreeNode], low: int, high: int) -> int:
        res = 0
        def traverse(node) -> None:
            if not node:
                return
            traverse(node.left)
            if low <= node.val <= high:
                nonlocal res
                res += node.val
            traverse(node.right)
        traverse(root)
        return res

    def DFS(self, root: Optional[TreeNode], low: int, high: int) -> int:
        res = 0
        def recurse(node):
            if not node:
                return
            nonlocal res
            if low <= node.val <= high:
                res += node.val
            if low < node.val:
                recurse(node.left)
            if node.val < high:
                recurse(node.right)
        recurse(root)
        return res

File: leet_BST_0938_rangesum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func rangeSumBST(root *TreeNode, low int, high int) int {
    Solutions := []func(* TreeNode, int, int) int {
        inorder,
        // BFS,
        // DFS,
    }
    return Solutions[ 0 ]( root, low, high )
}

func inorder(root *TreeNode, low int, high int) int {

    res := 0
    var traverse func( * TreeNode )
    traverse = func(node * TreeNode) {
        if node == nil {
            return
        }
        traverse(node.Left)
        n := node.Val
        if low <= n && n <= high {
            res += n
        }
        traverse(node.Right)
    }
    traverse(root)
    return res
}

func BFS(root *TreeNode, low int, high int) int {
    Q := [] *TreeNode { root }
    arr := []int{}
    for len(Q) > 0 {
        node := Q[0]
        Q = Q[1:]
        n := node.Val
        if low <= n && n <= high {
            arr = append(arr, n)
        }
        if node.Left != nil {
            Q = append(Q, node.Left)
        }
        if node.Right != nil {
            Q = append(Q, node.Right)
        }
    }
    res := 0
    for _, n := range arr {
        res += n
    }
    return res
}

func DFS (root * TreeNode, low int, high int) int {
    res := 0
    var recurse func( * TreeNode )
    recurse = func(node * TreeNode) {
        if node == nil {
            return
        }
        n := node.Val
        if low <= n && n <= high {
            res += n
        }
        if low < n {
            recurse(node.Left)
        }
        if n < high {
            recurse(node.Right)
        }
    }
    recurse(root)
    return res
}

